src/Disks/DiskCache.cpp

```diff
@@ -0,0 +1,1017 @@
+    }
+    void readStrict(char & c) override
+    {
+        from->readStrict(c);
+    }
+    size_t read(char * to, size_t n) override
+    {
+        return from->read(to, n);
+    }
+    void readStrict(char * to, size_t n) override
+    {
+        from->readStrict(to, n);
+    }
+    void prefetch() override
+    {
+        from->prefetch();
+    }
+
+private:
+    void finalize()
+    {
+        cache->finalize();
+        cache.reset(nullptr);
+
+        if (cache_part_size > expected_size)
+        {
+            LOG_ERROR(log, "Unexpected part size, got {}, expected {}", cache_part_size, expected_size);
+            throw Exception("Unexpected part size", ErrorCodes::LOCAL_CACHE_ERROR);
+        }
+
+        cache_disk->moveFile(cache_path + ".tmp", cache_path);
+        LOG_TRACE(log, "Add key {}: {}+{} of {}", cache_key, part_offset, cache_part_size, cache_file_size);
+        auto to_remove = cache_policy->add(cache_key, cache_file_size, part_offset, cache_part_size, false);
+        for (auto & entry_to_remove : to_remove)
+        {
+            String remove_cache_path = DiskCache::getCacheBasePath(entry_to_remove.first) + "_" + std::to_string(entry_to_remove.second);
+            LOG_TRACE(log, "Remove cached file {}", remove_cache_path);
+            cache_disk->removeFileIfExists(remove_cache_path);
+        }
+    }
+
+    void clean()
+    {
+        cache.reset(nullptr);
+        cache_policy->error(cache_key, part_offset);
+        cache_disk->removeFileIfExists(cache_path + ".tmp");
+    }
+
+    std::unique_ptr<ReadBuffer> from;
+    std::unique_ptr<WriteBufferFromFileBase> cache;
+    std::shared_ptr<DiskLocal> cache_disk;
+    std::shared_ptr<DiskCachePolicy> cache_policy;
+    String cache_key;
+    String cache_path;
+    // total size on remote FS
+    size_t cache_file_size = 0;
+    size_t part_offset = 0;
+    size_t expected_size = 0;
+
+    size_t cache_part_size = 0;
+
+    Poco::Logger * log = nullptr;
+};
+
+
+class CacheableMultipartReadBufferDecorator : public ReadBuffer
+{
+public:
+    CacheableMultipartReadBufferDecorator(std::shared_ptr<DiskLocal> cache_disk_, std::shared_ptr<DiskCachePolicy> cache_policy_,
+        const String & cache_key_, const String & cache_base_path_,
+        size_t offset_, size_t size_,
+        std::shared_ptr<DiskCacheDownloader> downloader_) :
+        ReadBuffer(nullptr, 0), cache_disk(std::move(cache_disk_)),
+        cache_policy(std::move(cache_policy_)), cache_key(cache_key_), cache_base_path(cache_base_path_),
+        current_offset(offset_), end_offset(size_ ? offset_ + size_ : size_),
+        downloader(std::move(downloader_))
+    {
+        ProfileEvents::increment(ProfileEvents::DiskCacheRequestsIn, 1);
+        log = &Poco::Logger::get("CacheableMultipartReadBufferDecorator");
+        ensureFrom();
+    }
+
+    void set(BufferBase::Position, size_t, size_t) override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
```

Comment:
btw I think this method needs to be implemented also because it is used in case when asynchronous reads are turned on (in ReadBufferFromS3 right after impl is initialized under flag use_external_buffer).

Reply:
Done.